File: src/main/java/io/odxproxy/OdxProxy.java
package io.odxproxy;

import com.fasterxml.jackson.core.type.TypeReference;
import com.github.f4b6a3.ulid.UlidCreator;
import io.odxproxy.client.OdxProxyClient;
import io.odxproxy.client.OdxProxyClientInfo;
import io.odxproxy.model.OdxClientKeywordRequest;
import io.odxproxy.model.OdxClientRequest;
import io.odxproxy.model.OdxServerResponse;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Main entry point for interacting with the ODX Proxy.
 * Provides static methods for common Odoo operations.
 */
public final class OdxProxy {

    private OdxProxy() {} // Prevent instantiation

    /**
     * Initializes the OdxProxyClient singleton with the required configuration.
     * This must be called once before any other methods are used.
     *
     * @param options The configuration options.
     * @return The initialized OdxProxyClient instance.
     */
    public static OdxProxyClient init(OdxProxyClientInfo options) {
        return OdxProxyClient.init(options);
    }

    private static OdxProxyClient client() {
        return OdxProxyClient.getInstance();
    }
    
    private static OdxClientKeywordRequest copyKeywords(OdxClientKeywordRequest keyword) {
        return new OdxClientKeywordRequest(keyword);
    }

    /**
     * Performs a search and returns matching record IDs.
     */
    public static CompletableFuture<OdxServerResponse<List<Integer>>> search(String model, List<Object> params, OdxClientKeywordRequest keyword, String id) {
        OdxClientKeywordRequest kCopy = copyKeywords(keyword);
        kCopy.sort = null;
        kCopy.limit = null;
        kCopy.offset = null;
        kCopy.fields = null;

        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "search";
        body.modelId = model;
        body.keyword = kCopy;
        body.params = params;
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<List<Integer>>>() {});
    }
    
    /**
     * Performs a search and reads the data of matching records.
     */
    public static <T> CompletableFuture<OdxServerResponse<List<T>>> searchRead(String model, List<Object> params, OdxClientKeywordRequest keyword, String id, Class<T> resultType) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "search_read";
        body.modelId = model;
        body.keyword = copyKeywords(keyword);
        body.params = params;
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<List<T>>>() {});
    }

    /**
     * Reads the data for a specific set of record IDs.
     */
    public static <T> CompletableFuture<OdxServerResponse<List<T>>> read(String model, List<Integer> ids, OdxClientKeywordRequest keyword, String id, Class<T> resultType) {
        OdxClientKeywordRequest kCopy = copyKeywords(keyword);
        kCopy.sort = null;
        kCopy.limit = null;
        kCopy.offset = null;

        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "read";
        body.modelId = model;
        body.keyword = kCopy;
        body.params = Collections.singletonList(ids);
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<List<T>>>() {});
    }

    /**
     * Retrieves metadata for the fields of a model.
     */
     public static <T> CompletableFuture<OdxServerResponse<T>> fieldsGet(String model, OdxClientKeywordRequest keyword, String id) {
        OdxClientKeywordRequest kCopy = copyKeywords(keyword);
        kCopy.sort = null;
        kCopy.limit = null;
        kCopy.offset = null;
        kCopy.fields = null;
        
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "fields_get";
        body.modelId = model;
        body.keyword = kCopy;
        body.params = Collections.emptyList();
        body.odooInstance = client().getOdooInstance();
        
        return client().postRequest(body, new TypeReference<OdxServerResponse<T>>() {});
    }

    /**
     * Counts the number of records matching the search criteria.
     */
    public static CompletableFuture<OdxServerResponse<Integer>> searchCount(String model, List<Object> params, OdxClientKeywordRequest keyword, String id) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "search_count";
        body.modelId = model;
        body.keyword = copyKeywords(keyword);
        body.params = params;
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<Integer>>() {});
    }

    /**
     * Creates a new record.
     */
    public static <T> CompletableFuture<OdxServerResponse<T>> create(String model, List<Object> params, OdxClientKeywordRequest keyword, String id) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "create";
        body.modelId = model;
        body.keyword = copyKeywords(keyword);
        body.params = params;
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<T>>() {});
    }

    /**
     * Updates existing records.
     */
    public static CompletableFuture<OdxServerResponse<Boolean>> write(String model, List<Integer> ids, Object values, OdxClientKeywordRequest keyword, String id) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "write";
        body.modelId = model;
        body.keyword = copyKeywords(keyword);
        body.params = Arrays.asList(ids, values);
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<Boolean>>() {});
    }
    
    /**
     * Deletes records by their IDs.
     */
    public static CompletableFuture<OdxServerResponse<Boolean>> remove(String model, List<Integer> ids, OdxClientKeywordRequest keyword, String id) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "unlink";
        body.modelId = model;
        body.keyword = copyKeywords(keyword);
        body.params = Collections.singletonList(ids);
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<Boolean>>() {});
    }
    
    /**
     * Calls an arbitrary method on a model. On some action you may need to subclass OdxClientKeywordRequest to add extra fields required by the method.
     */
    public static <T> CompletableFuture<OdxServerResponse<T>> callMethod(String model, String functionName, List<Object> params, OdxClientKeywordRequest keyword, String id) {
        OdxClientRequest body = new OdxClientRequest();
        body.id = Optional.ofNullable(id).orElseGet(() -> UlidCreator.getUlid().toString());
        body.action = "call_method";
        body.modelId = model;
        body.fnName = functionName;
        body.keyword = copyKeywords(keyword);
        body.params = params;
        body.odooInstance = client().getOdooInstance();

        return client().postRequest(body, new TypeReference<OdxServerResponse<T>>() {});
    }
}

File: src/main/kotlin/io/odxproxy/OdxProxy.kt
package io.odxproxy

import com.github.f4b6a3.ulid.UlidCreator
import io.odxproxy.client.OdxProxyClient
import io.odxproxy.client.OdxProxyClientInfo
import io.odxproxy.model.OdxClientKeywordRequest
import io.odxproxy.model.OdxClientRequest
import io.odxproxy.model.OdxServerResponse
import io.odxproxy.model.toJsonElement
import kotlinx.serialization.json.JsonArray
import java.util.concurrent.CompletableFuture

public object OdxProxy {
    private fun generateId(providedId: String?): String {
        return providedId ?: UlidCreator.getUlid().toString()
    }
    private fun client(): OdxProxyClient = OdxProxyClient.getInstance()

    @JvmStatic
    public fun init(options: OdxProxyClientInfo): OdxProxyClient {
        return OdxProxyClient.init(options)
    }

    @JvmStatic
    public fun search(
        model: String,
        params: List<Any?>,
        keyword: OdxClientKeywordRequest,
        id: String?
    ): CompletableFuture<OdxServerResponse<List<Int>>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "search",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = toJsonElement(params),
            odooInstance = client().odooInstance
        )
        return client().postRequestList(request, Int::class.javaObjectType)
    }

    @JvmStatic
    public fun <T> searchRead(
        model: String,
        params: List<Any?>,
        keyword: OdxClientKeywordRequest,
        id: String?,
        resultType: Class<T>
    ): CompletableFuture<OdxServerResponse<List<T>>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "search_read",
            modelId = model,
            keyword = keyword,
            params = toJsonElement(params),
            odooInstance = client().odooInstance
        )
        return client().postRequestList(request, resultType)
    }

    @JvmStatic
    public fun <T> read(
        model: String,
        ids: List<Int>,
        keyword: OdxClientKeywordRequest,
        id: String?,
        resultType: Class<T>
    ): CompletableFuture<OdxServerResponse<List<T>>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "read",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = toJsonElement(listOf(ids)),
            odooInstance = client().odooInstance
        )
        return client().postRequestList(request, resultType)
    }

    @JvmStatic
    public fun searchCount(
        model: String,
        params: List<Any?>,
        keyword: OdxClientKeywordRequest,
        id: String?
    ): CompletableFuture<OdxServerResponse<Int>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "search_count",
            modelId = model,
            keyword = keyword,
            params = toJsonElement(params),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, Int::class.javaObjectType)
    }

    @JvmStatic
    public fun <T> create(
        model: String,
        params: List<Any?>,
        keyword: OdxClientKeywordRequest,
        id: String?,
        resultType: Class<T>
    ): CompletableFuture<OdxServerResponse<T>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "create",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = toJsonElement(params),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, resultType)
    }

    @JvmStatic
    public fun write(
        model: String,
        ids: List<Int>,
        values: Any,
        keyword: OdxClientKeywordRequest,
        id: String?
    ): CompletableFuture<OdxServerResponse<Boolean>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "write",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = toJsonElement(listOf(ids, values)),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, Boolean::class.javaObjectType)
    }

    @JvmStatic
    public fun remove(
        model: String,
        ids: List<Int>,
        keyword: OdxClientKeywordRequest,
        id: String?
    ): CompletableFuture<OdxServerResponse<Boolean>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "unlink",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = toJsonElement(listOf(ids)),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, Boolean::class.javaObjectType)
    }

    @JvmStatic
    public fun <T> fieldsGet(
        model: String,
        keyword: OdxClientKeywordRequest,
        id: String?,
        resultType: Class<T>
    ): CompletableFuture<OdxServerResponse<T>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "fields_get",
            modelId = model,
            keyword = keyword.resetPagination(),
            params = JsonArray(emptyList()),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, resultType)
    }

    @JvmStatic
    public fun <T> callMethod(
        model: String,
        functionName: String,
        params: List<Any?>,
        keyword: OdxClientKeywordRequest,
        id: String?,
        resultType: Class<T>
    ): CompletableFuture<OdxServerResponse<T>> {
        val request = OdxClientRequest(
            id = generateId(id),
            action = "call_method",
            modelId = model,
            fnName = functionName,
            keyword = keyword,
            params = toJsonElement(params),
            odooInstance = client().odooInstance
        )
        return client().postRequest(request, resultType)
    }
}